src/transport/PeerCache.h

```diff
@@ -0,0 +1,187 @@
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <core/CHIPError.h>
+#include <core/ReferenceCounted.h>
+#include <support/CodeUtils.h>
+#include <support/Pool.h>
+#include <support/ReferenceCountedHandle.h>
+#include <support/logging/CHIPLogging.h>
+#include <system/TimeSource.h>
+#include <transport/AdminPairingTable.h>
+#include <transport/raw/PeerAddress.h>
+
+namespace chip {
+namespace Transport {
+
+class PeerCacheEntry;
+using PeerHandle = ReferenceCountedHandle<PeerCacheEntry>;
+
+class PeerCacheEntryDeleter
+{
+public:
+    // This is a no-op because life-cycle of PeerCache is rotated by LRU
+    static void Release(PeerCacheEntry * entry) {}
+};
+
+/**
+ * @brief
+ *   A Peer table entry stores the binding of NodeId, TransportAddress, and message counters. For both secure and unsecure nodes.
+ *
+ *   The entries are rotated using LRU, but entry can be hold by using PeerHandle, which increase the reference count by 1. When the
+ * reference count is not 0, the entry won't be pruned.
+ */
+class PeerCacheEntry : public ReferenceCounted<PeerCacheEntry, PeerCacheEntryDeleter>
+{
+public:
+    PeerCacheEntry(NodeId peerNodeId, Transport::AdminId admin) : mPeerNodeId(peerNodeId), mAdmin(admin) {}
+
+    PeerCacheEntry(const PeerCacheEntry &) = delete;
+    PeerCacheEntry & operator=(const PeerCacheEntry &) = delete;
+    PeerCacheEntry(PeerCacheEntry &&)                  = delete;
+    PeerCacheEntry & operator=(PeerCacheEntry &&) = delete;
+
+    NodeId GetPeerNodeId() const { return mPeerNodeId; }
+    Transport::AdminId GetAdminId() const { return mAdmin; }
+
+    uint64_t GetLastActivityTimeMs() const { return mLastActivityTimeMs; }
+    void SetLastActivityTimeMs(uint64_t value) { mLastActivityTimeMs = value; }
+
+    const PeerAddress & GetPeerAddress() const { return mPeerAddress; }
+    PeerAddress & GetPeerAddress() { return mPeerAddress; }
+    void SetPeerAddress(const PeerAddress & address) { mPeerAddress = address; }
+
+    /// @brief Hold a cache entry to prevent it from being released.
+    PeerHandle Hold() { return PeerHandle(this); }
+
+private:
+    uint64_t mLastActivityTimeMs = 0;
+
+    const NodeId mPeerNodeId;
+    const Transport::AdminId mAdmin;
+
+    PeerAddress mPeerAddress = PeerAddress::Uninitialized();
+};
+
+template <size_t kMaxConnectionCount, Time::Source kTimeSource = Time::Source::kSystem>
+class PeerCache
+{
+public:
+    /**
+     * Allocates a new peer state object out of the internal resource pool.
+     *
+     * @returns CHIP_NO_ERROR if state could be initialized. May fail if maximum connection count
+     *          has been reached (with CHIP_ERROR_NO_MEMORY).
+     */
+    CHECK_RETURN_VALUE
+    CHIP_ERROR AllocEntry(NodeId peerNodeId, Transport::AdminId admin, PeerCacheEntry *& peer)
```

Comment:
NodeID is not unique across peers. 

Spec says that uniqueness is guaranteed only by a triplet of 'rootpk(hash), fabricid, nodeid'